lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,506 @@
+//     minimum network fee.
+//  6. Compute the congestion threshold as congestionFactorBasisPoints * maxSizeOfNumBlocks / 10000.
+//  7. If the total size of the transactions in the slice is less than the congestion threshold, return one bucket lower
+//     than the priority fee bucket from step 4.
+//  8. If the total size of the transactions in the size is greater than the congestion threshold and less
+//     than or equal to maxSizeOfNumBlocks, return the priority fee bucket from step 4.
+//  9. Otherwise, return one bucket higher than the priority fee bucket from step 4.
+func (posFeeEstimator *PoSFeeEstimator) estimateFeeRateNanosPerKBGivenTransactionRegister(
+	txnRegister *TransactionRegister,
+	congestionFactorBasisPoints uint64,
+	priorityPercentileBasisPoints uint64,
+	maxBlockSize uint64,
+) (uint64, error) {
+	txnRegister.RLock()
+	defer txnRegister.RUnlock()
+	it := txnRegister.GetFeeTimeIterator()
+	maxSizeOfNumBlocks := maxBlockSize * posFeeEstimator.numBlocks
+	totalTxnsSize := uint64(0)
+	var txns []*MempoolTx
+	for it.Next() {
+		tx, ok := it.Value()
+		if !ok {
+			break
+		}
+		txnBytes, err := tx.Tx.ToBytes(false)
+		if err != nil {
+			return 0, errors.Wrap(err, "estimateFeeRateNanosPerKBGivenTransactionRegister: Problem serializing txn")
+		}
+		totalTxnsSize += uint64(len(txnBytes))
+		txns = append(txns, tx)
+		// TODO: I think we want to include the txn that puts us over the limit, but
+		// we can just move this check up a few lines if that's wrong.
+		if totalTxnsSize > maxSizeOfNumBlocks {
+			break
+		}
+	}
+	globalMinFeeRate, _ := txnRegister.minimumNetworkFeeNanosPerKB.Uint64()
+	if len(txns) == 0 {
+		// If there are no txns in the transaction register, we simply return the minimum network fee.
+		return globalMinFeeRate, nil
+	}
+
+	bucketMinFee, bucketMaxFee, err := getPriorityFeeBucketFromTxns(
+		txns,
+		priorityPercentileBasisPoints,
+		txnRegister.minimumNetworkFeeNanosPerKB,
+		txnRegister.feeBucketGrowthRateBasisPoints)
+	if err != nil {
+		return 0, errors.Wrap(err,
+			"estimateFeeRateNanosPerKBGivenTransactionRegister: Problem computing priority fee bucket")
+	}
+	// If the bucketMinFee is less than or equal to the global min fee rate, we return the global min fee rate.
+	if bucketMinFee <= globalMinFeeRate {
+		return globalMinFeeRate, nil
+	}
+
+	// Compute the congestion threshold. If our congestion factor is 100% (or 10,000 bps),
+	// then congestion threshold is simply max block size * numBlocks
+	// TODO: I don't know if I like this name really.
+	congestionThreshold := (congestionFactorBasisPoints * maxSizeOfNumBlocks) / MaxBasisPoints
+	// If the total size of the txns in the transaction register is less than the computed congestion threshold,
+	// we return one bucket lower than the Priority fee.
+	if totalTxnsSize <= congestionThreshold {
+		// Return one bucket lower than Priority fee
+		oneBucketLowerMinFee, _ := computeFeeTimeBucketRangeFromFeeNanosPerKB(
```

Comment:
I am a little bit worried that you'll run into a rounding error here because you're going: fee -> bucket exponent -> minFee -> subtract one -> back to bucket exponent. The reason is there could be an off by one in the first "bucket exponent -> minFee" step. Better would be to go: fee -> bucket exponent -> bucket exponent - 1 -> minFee.